Frontier silicon

zhelev opened the pull request:
**Description:**
new component - frontier silicon, 'depends on latest' netdisco, already online.
This a second attempt to pull these changes.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#1966

**Example entry for `configuration.yaml` (if applicable):**
```yaml
media_player:
  - platform: frontier_silicon
    host: IP_ADDRESS
```
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

zhelev committed:
added frontier_silicon constant

zhelev committed:
added frontier_silicon constant

zhelev committed:
added frontier_silicon constant

zhelev committed:
added the frontier_silicon component

zhelev committed:
added the frontier_silicon component

zhelev committed:
added the frontier_silicon component

zhelev committed:
added the frontier_silicon component

zhelev committed:
added the frontier_silicon component

zhelev committed:
added the frontier_silicon component

zhelev committed:
cleaning up according to travis

zhelev committed:
cleaning up according to travis

zhelev committed:
cleaning up according to travis

zhelev committed:
cleaning up according to travis

zhelev committed:
trying to satisfy pylint

zhelev committed:
trying to satisfy pylint

zhelev committed:
trying to satisfy pylint

zhelev committed:
trying to satisfy pylint

zhelev committed:
trying to satisfy pylint

zhelev committed:
trying to satisfy pylint

zhelev committed:
trying to satisfy pylint

zhelev committed:
trying to satisfy pylint

zhelev committed:
trying to satisfy pylint

zhelev committed:
trying to satisfy pylint

zhelev committed:
fsapi version 0.0.6

zhelev committed:
fsapi version 0.0.6

zhelev committed:
fsapi version 0.0.6

zhelev committed:
fsapi version 0.0.6

zhelev committed:
fsapi version 0.0.6

zhelev committed:
fsapi version 0.0.6

zhelev committed:
with fsapi version 0.0.7

zhelev committed:
with fsapi version 0.0.7

zhelev committed:
with fsapi version 0.0.7

zhelev committed:
with fsapi version 0.0.7

zhelev committed:
added fsapi dependency

zhelev committed:
added fsapi dependency

zhelev committed:
added fsapi dependency

zhelev committed:
added fsapi dependency

zhelev committed:
yielding the FSAPI

zhelev committed:
yielding the FSAPI

zhelev committed:
yielding the FSAPI

zhelev committed:
yielding the FSAPI

zhelev committed:
Removing white space from docstring

zhelev committed:
Removing white space from docstring

zhelev committed:
Removing white space from docstring

zhelev committed:
Removing white space from docstring

zhelev committed:
Removing white space from an empty line

zhelev committed:
Removing white space from an empty line

zhelev committed:
Removing white space from an empty line

zhelev committed:
Removing white space from an empty line

zhelev committed:
Switching to sync

zhelev committed:
Switching to sync

zhelev committed:
Switching to sync

zhelev committed:
Switching to sync

zhelev committed:
clean up white spaces and rename device to FSAPIDevice

zhelev committed:
clean up white spaces and rename device to FSAPIDevice

zhelev committed:
clean up white spaces and rename device to FSAPIDevice

zhelev committed:
clean up white spaces and rename device to FSAPIDevice

zhelev committed:
changed info to debug

zhelev committed:
changed info to debug

zhelev committed:
Merge branch 'frontier_silicon' of https://github.com/zhelev/home-assistant into frontier_silicon

zhelev committed:
merging

zhelev committed:
remove white space

zhelev committed:
generated requirements

zhelev committed:
generated requirements

zhelev committed:
Merge branch 'dev' into frontier_silicon

zhelev committed:
rere-merged files

zhelev committed:
pylint

zhelev committed:
moved import requests to the method where it is being used

zhelev committed:
add a basic unit test

zhelev committed:
cleaned up source code

zhelev committed:
added frontier_silicon constant

zhelev committed:
added the frontier_silicon component

zhelev committed:
added basic test

zhelev committed:
added fsapi to requirements_all.txt

zhelev committed:
added coverage omit, though a basic test was included

armills commented on homeassistant/components/media_player/frontier_silicon.py:
```diff
@@ -0,0 +1,245 @@
+        info_name = fs_device.play_info_name
+        info_text = fs_device.play_info_text
+
+        if artist:
```
Don't stuff everything into the title. Use the additional media fields as found here: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/media_player/__init__.py#L478

zhelev replied:
Thanks, somehow I missed those. The platform handles artist and album name now, though album name does not show in the UI. I've added the music media type so the artist name will be rendered.

armills commented on homeassistant/components/media_player/frontier_silicon.py:
```diff
@@ -0,0 +1,245 @@
+        """Create a fsapi session."""
+        from fsapi import FSAPI
+
+        return FSAPI(self._device_url, self._password)
```
Is there a reason you don't create this object in `__init__`, and keep the reference? We don't want to recreate it each time it needs to be used.

zhelev replied:
Hi, thanks for the input. I do appreciate it. 
There is a single session controlling a frontier silicon device. Once a new connection has been established - the previous session will be invalidated. This makes sense as an Internet Radio can play a single station, song, etc at a time. The reason I always recreate the reference when 'update' is being called, is to ensure I have a valid session. Otherwise I would always have to try-catch-recreate in case someone has used the the Undock app (i.e firmware update, etc.) in between and has destroyed the __init__ session. I am not a Python developer but usually a http request (once every 30 second in the local network) does not cause much overhead. In my experience this pattern is much more robust and  maintainable than posting a request, catching an exception, then eventually recreating the instance and repeating the request. Still I am not a Python developer and I am sure this not the first platform in homeassistant handling this issue. If you have any suggestions, please comment here. I've used the KISS solution here. I would implement an async component as soon as I find time, the fsapi is the bottleneck in this case but I 'll fix it as soon as I can.

armills replied:
I'm OK with that behavior then. Just a few more thoughts:

1) Can you include a short version of this explanation in the function docstring? That way there is an explanation showing that this is being done by design.

2) You should consider making this a property for readability. That way we can document in one place why it is recreated for each request, but we don't have to get bogged down with the details at each usage. The preference in Python is to use properties in favor of get_* method patterns.

```python
@property
def fs_device(self):
    """Create a fsapi session.

    Explanation text here ...
    """
    from fsapi import FSAPI

    return FSAPI(self._device_url, self._password)

# Example:
def media_play(self):
    """Send play command."""
    self.fs_device.play()
```

emlove reviewed (changes requested):
Looks really good overall. Just a few comments.

zhelev committed:
Merge branch 'frontier_silicon' of https://github.com/zhelev/home-assistant into frontier_silicon

zhelev committed:
added MEDIA_TYPE_MUSIC for artist and album

zhelev committed:
Merge branch 'dev' into frontier_silicon

zhelev committed:
removed duplicate cons

armills commented on homeassistant/components/media_player/frontier_silicon.py:
```diff
@@ -0,0 +1,268 @@
+            self._album_name = album
+
+        # album name not shown in the UI, quick dirty fix
+        # self._title = album + ' - ' + title
```
The commented code needs to be removed. We can post quick and dirty fixes outside of the repo.

armills commented on homeassistant/components/media_player/frontier_silicon.py:
```diff
@@ -0,0 +1,268 @@
+
+        self._title = title
+
+        self._artist = ''
```
This can just be `self._artist = fs_device.play_info_artist`. Hass will treat media fields that return `None` as being not specified. Same for album_name below.

armills commented on homeassistant/components/media_player/frontier_silicon.py:
```diff
@@ -0,0 +1,268 @@
+                title += ' - '
+            title += info_name
+        if info_text:
+            title += ': ' + info_text
```
I think this could be reduced to the following code. That way the ': ' will only be used if both info_name and info_text are set. Also I think the ' - ' was leftover from when more fields were being concatenated. Alternatively you might want to just use the info_name, but that's your call since I'm not sure what's returned by the platform lib.

string.join will join a list of strings together with the given string as the separator. filter removes matching values from the list.

```python
self._title = ': '.join(filter(None, [info_name, info_text]))
```

zhelev committed:
switched fsapi call to a property, removed unecessary comment

zhelev committed:
detailed docstring for fs_device

armills commented on homeassistant/components/media_player/frontier_silicon.py:
```diff
@@ -0,0 +1,254 @@
+        info_name = fs_device.play_info_name
+        info_text = fs_device.play_info_text
+
+        self._title = '- '.join(filter(None, [info_name, info_text]))
```
If the dash will be the separator we should have a space on both sides, so `' - '.join(`

armills commented on homeassistant/components/media_player/frontier_silicon.py:
```diff
@@ -0,0 +1,254 @@
+    def volume_up(self):
+        """Send volume up command."""
+        fs_device = self.fs_device
+        fs_device.volume = (fs_device.volume + 1)
```
If you change this to `self.fs_device.volume += 1`, the property will only be accessed once without requiring the intermediate variable. Same below.

emlove reviewed (commented):
Just a few really small adjustments. Otherwise this is in good shape.

homeassistant commented:
Hello @zhelev,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`9978915f800c345adb2d521a114ba1a31fe1b138`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;6131&#x2F;commits&#x2F;9978915f800c345adb2d521a114ba1a31fe1b138) - This commit has something that looks like an email address (unrz205@htpc). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hello @zhelev,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`9978915f800c345adb2d521a114ba1a31fe1b138`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;6131&#x2F;commits&#x2F;9978915f800c345adb2d521a114ba1a31fe1b138) - This commit has something that looks like an email address (unrz205@htpc). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

zhelev committed:
added a space for the info_name - info_text separator

zhelev committed:
reduced proeprty (fsapi) access for volume down/up

zhelev commented:
I've fixed the issues. Unfortunatelly I've commited once with a wrong user - I am currently abroad and developing on my mobile. I've reverted the changes and the git tree looks good but the cla-error seems to persist. Maybe it will disappear after my build completes which seems to be stuck. The Python 3.6 build is waiting to start for 3 and a halfs hours now. I've contacted the travis guys, hopefully they will have a look soon, though it is Fasching today in Germany and this could take some time. Thanks for the comments, appreciate it, have a nice day.

zhelev commented:
Just for info - the response from travis:
_I've looked through our logs and I'm unfortunately not sure what happened here. To find out if the issue is constant, I've restarted your build. Take note that I've given your organization a +5 additional capacity so that this operation doesn't impact your productivity._

emlove reviewed (approved):
Awesome! Thanks for following up on all the feeback!
